03-task-manager/optionalArrayMethodsReviewExtraAssignment.js

```diff
@@ -0,0 +1,291 @@
+    this.data++;
+    console.log('OOP demo:', this.data);
+  }
+}
+
+// Every time we call `.behavior()`, the data (number) inside `object` is
+// incremented by 1, so we print "OOP demo 1", "OOP demo 2", etc.
+object.behavior()
+object.behavior()
+object.behavior()
+object.behavior()
+object.behavior()
+
+// That's why we call them "array methods" they are methods that exist on the
+// "Array" object.
+
+/////////////////////////////////////////////////
+///// What are the common JS array methods? /////
+//
+// - Array.prototype.push (does not take a callback)
+//
+// These guys all take a callback as input, and then call the callabck for
+// each item in the array
+//
+// - Array.prototype.filter
+//   - The callback should return a boolean. If the return value is true, the
+//     element becomes a member of the new array. If the return value is false,
+//     the element is filtered (removed).
+   const integers = [1, 2, 3, 4, 5];
+   // evenNumbers will be interger % 2 for each integer
+   const evenNumbers = integers.filter((integer) => {
+     return integer % 2 === 0
+   })
+
+// - Array.prototype.map
+//   - The callback recieves each item of the array. The return value is pushed
+//     into a new array
+   const numbers = [1, 2, 3];
+   const doubles = numbers.map((i) => i * 2);
+
+// - Array.prototype.forEach
+//   - `forEach` is like a "for" loop. It calls the callback for every item in
+//     the array
+   evenNumbers.forEach((thingy) => console.log('even', thingy));
+   doubles.forEach((d) => console.log('doubled!', d));
+
+// - Array.prototype.reduce
+//   - Reduce is weird & tricky
+//   - We ain't going there right now
+
+/////////////////////////// CHALLENGES ////////////////////////////////////////
+
+// Each challenge will be related to this array of names. It will pose a
+// problem related to these names, and then implement the solution. The
+// challenges are:
+//
+// - Create a new array with only each person's last name
+// - Filter names that don't match the format "<first> <last>"
+//   - Should remove Tam because she has a double-space
+//   - Should remove Carlow because he has a middle-name
+//   - Should also remove names like:
+//     - "Timothy      Cook"
+//     - "Nick_Masters"
+//     - "Timmy-Turner"
+//     - "Billy\nBob"
+//     - etc.
+// - Create a new array where everyone's name is converted to "Title Case"
+//   - The first character of each word should be uppercase
+//   - All other characters in the word should be lowercase
+//   - expected output is ['Dimitry Santiago', 'Carlos D. Perez', 'Tam Person', ...]
+// - Last Challenge:
+//     Remove names with the wrong format
+//     AND change it to "Title Case"
+//     AND remove people whose last name ends with z
+//     AND write a message asking them to sign up
+//
+// For an extra assignment, you may implement these yourself send a message
+// in the class discussion slack channel to ask for submission instructions.
```

Comment:
I think, instead of having each student ask how to submit the the assignment, we'd just suggest that they add their changes to this file and include them in their PR.

Reply:
OK, I've done that. Let's also add a note to week 3 reviewer notes for reviewers to beware that this may be included.